### DISCUSSION TITLE
penmod1.2.0修改版安装脚本

### DISCUSSION THREAD

**Discussion by tofuture-1** [PenUniverse/PenMods#177] [奇技淫巧]
# 背景
根据PenMods作者于社区发布的贴子，由于原社区数据丢失，目前使用原版安装程序无法正常安装PenMods。
因此，在PenMods1.3.0更新并修复正常安装功能前，此帖提供一个基于原版installer-1.1.0安装脚本修改的安装脚本及相关文件[https://github.com/tofuture-1/penmod_install_for1.2.0]，专用于有道2，2.0.7/2.1.2系统版本的PenMods1.2.0安装。
# 使用方法
下载链接中的releases文件，解压后使用任意python3.9及以上的IDE打开并运行该文件夹中的installer.py。
![批注 2024-02-03 233850](https://github.com/PenUniverse/PenMods-release/assets/80562703/241c4262-d5e3-4f9c-b9b1-962243d47bad)

**Comment by JohnXX123456** [comment]
请问程序显示当前系统已安装PenMods，但词典笔界面没有变化是什么原因

**Reply by tofuture-1** [reply]
设置有变化吗，没有的话清数据再装试试

**Reply by JohnXX123456** [reply]
我开始想装ydpExt没成功，然后就把它在userdisk里的文件删了，可能是因为删错了，重启后就没PenMods，然后系统就从2.0.7升到2.1.2，我把系统装回1.6后又装回2.0.7，连网后过一会还是会自己升回2.1.2。我开始自己升到过2.1.2，第一次装penmods 的时候刷了2.0.7系统，也是一打开联网就开始下载然后安装，我开始准备等到1.30支持2.1.2再装penmods, 不过放了几天后又试了一次就稀里糊涂的装上了。现在userdisk里只有6个文件（之前装上penmods时有10个）
![屏幕截图 2024-02-21 194025](https://github.com/PenUniverse/PenMods-release/assets/160201935/293abfd2-b801-467a-8c12-007758377fb1)

**Reply by tofuture-1** [reply]
目前penmod能正常用吗？

**Reply by JohnXX123456** [reply]
不能，没有penmod，
但在有道界面能正常用

**Reply by tofuture-1** [reply]
那你试试重新刷一遍2.0.7，开机连手机热点，一激活马上关掉热点，后面再按正常流程刷一遍penmod

**Reply by JohnXX123456** [reply]
刷完penmods还是没变化，再刷显示当前系统已安装PenMods
![屏幕截图 2024-02-22 185115](https://github.com/PenUniverse/PenMods-release/assets/160201935/499b6210-0c3d-46b9-a4b8-b9463d1db1b1)

**Reply by tofuture-1** [reply]
这次刷之前有重新刷系统吗？

**Reply by JohnXX123456** [reply]
我会207后又刷的penmods,第一次刷完后没变化，再刷显示当前系统已安装PenMods

**Reply by tofuture-1** [reply]
你可以试试2.1.2直接装，之前原作者说过2.1.2没什么更改，应该是可以适用，这个帖子的修改版脚本我记得因为获取不了适配文件也把版本对应判定注释掉了

**Reply by tofuture-1** [reply]
或者加讨论组问问其他人有没有这种情况，之前也见过有人用原版安装程序装了没反应的，我暂时也没有别的办法能解决

**Reply by DSFdsfWxp** [reply]
看`Installer.py`212行。

```python
def isInstalled() -> bool:
    exec = adb.execute('shell cat /usr/bin/runDictPen')
    return exec.find('try_inject') == -1
```

自己去`/usr/bin/runDictPen`把`try_inject`字眼删干净就行了。

**Reply by tofuture-1** [reply]
真神归位了（手动doge）

**Reply by jdjfjdsfj** [reply]
> 看212行。`Installer.py`
> 
> ```python
> def isInstalled() -> bool:
>     exec = adb.execute('shell cat /usr/bin/runDictPen')
>     return exec.find('try_inject') == -1
> ```
> 
> 自己去把字眼删干净就行了。`/usr/bin/runDictPen``try_inject`
/usr/bin/runDictPen进不去怎么办

**Reply by chicken2-5** [reply]
我的一打开就闪退怎么回事，求解

**Reply by chicken2-5** [reply]
并且出现
Traceback (most recent call last):
  File "F:\Installer-main\Installer.py",line 2,in <module>
    import requests
MoudleNotFoundError:No module named 'requests'

**Reply by tofuture-1** [reply]
> 并且出现
> Traceback (most recent call last):
>   File "F:\Installer-main\Installer.py",line 2,in <module>
>     import requests
> MoudleNotFoundError:No module named 'requests'

看看Python版本是否满足要求

**Comment by MGtxzzg** [comment]
大家可以看看这是怎么回事吗，一运行打开adb调试就秒退，然后出现以下红色的字
Traceback (most recent call last):
  File "E:\Installer-1.1.0_revised_2\Installer-1.1.0(修改版)\Installer.py", line 302, in <module>
    sys_version = getSystemVersions()['Version']
  File "E:\Installer-1.1.0_revised_2\Installer-1.1.0(修改版)\Installer.py", line 193, in getSystemVersions
    exec = adb.execute('shell cat /Version')
  File "E:\Installer-1.1.0_revised_2\Installer-1.1.0(修改版)\Installer.py", line 84, in execute
    handleAssert(self.connect(), '请连接你的设备。')
  File "E:\Installer-1.1.0_revised_2\Installer-1.1.0(修改版)\Installer.py", line 126, in connect
    devices.append((dev[0],dev[1].removesuffix('\n')))
AttributeError: 'str' object has no attribute 'removesuffix'
>>>

**Reply by tofuture-1** [reply]
Python版本要3.9及以上，这里少了新增的功能了

**Reply by MGtxzzg** [reply]
好的谢谢！我换了一台电脑，然后我想把词典笔关机时，它自己开始了向2.1.2的更新，我就只好硬着头皮安装了penmods，没想到成功了，这不会有什么不良的后果吧
现在是在整newbing，但按照教程却一直不成功

**Reply by tofuture-1** [reply]
2.1.2按原作者的说法应该是也能兼容，没有太大区别

**Comment by waterycrane** [comment]
在vscode里打开的,有人知道为什么会这样吗
PS C:\Users\86159> & C:/Users/86159/AppData/Local/Programs/Python/Python311/python.exe "c:/Users/86159/Downloads/Installer-1.1.0_revised (1)/Installer-1.1.0(修改版)/Installer.py"
┌───────────────────────────────────┐
│                                   │
│       ╔═╗┌─┐┌┐┌╔╦╗┌─┐┌┬┐┌─┐       │
│       ╠═╝├┤ │││║║║│ │ ││└─┐       │
│       ╩  └─┘┘└┘╩ ╩└─┘─┴┘└─┘       │
│                                   │
└───────────────────────────────────┘
Welcome to use PenMods!
Developer:  RedbeanW
Repo:       https://github.com/PenUniverse/Installer
Version:    1.1.0
'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
无法获取软件包列表。
Assertion Failed: ADB服务是必须的。
'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
# 请将词典笔与电脑连接，并开启ADB调试。
# 请进入 [更多设置]->[关于]->[法律监管] 后快速连续点击5次屏幕即可打开ADB调试。
'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
正在等待...1s'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
正在等待...2s'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
正在等待...3s'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
正在等待...4s'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
正在等待...5s'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
正在等待...6s'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
正在等待...7s'adb' 不是内部或外部命令，也不是可运行的程序
或批处理文件。

**Reply by tofuture-1** [reply]
把Python脚本同目录文件夹里面的dependent/platform-tools/adb添加到系统环境变量试试
ChatGPT给的方法：
要在Windows系统中添加ADB工具的环境变量，请按照以下步骤进行操作：

1. 打开“此电脑”（或“计算机”）窗口，右键单击空白处，然后选择“属性”。
2. 在弹出的窗口中，点击“高级系统设置”链接。
3. 在系统属性窗口中，点击“环境变量”按钮。
4. 在“环境变量”窗口中，找到“系统变量”下的“Path”变量，选中它，然后点击“编辑”按钮。
5. 在弹出的编辑环境变量窗口中，点击“新建”，然后输入ADB工具所在的路径，点击“确定”。
   例如，如果ADB工具在C:\adb路径下，那么您应该添加一个新条目：C:\adb。
6. 确保已经保存了所有更改，并关闭所有打开的窗口。

**Reply by whoareyouwhathahha** [reply]
不管用啊

**Reply by tofuture-1** [reply]
你把python更新到最新，注意安装的时候勾选add to path

**Comment by laoaideqi** [comment]
G3词典笔怎么办，现在原程序用不了了，下这个修改版ui乱的

**Comment by Redbeanw44602** [comment]
现在可以去tg群里手动装1.3.0预览版

**Comment by AuroraAeon** [comment]
YDP021 装完以后一直卡在Logo页面咋办

**Reply by tofuture-1** [reply]
penmod本来就不适配ydp021，请在安装之前看看说明，该安装程序仅用于2.0.7及2.1.2的有道2安装。
卡logo你可以试试按重装有道满分系统的教程重装看，或者去tg群里问

**Reply by Redbeanw44602** [reply]
> penmod本来就不适配ydp021，请在安装之前看看说明，该安装程序仅用于2.0.7及2.1.2的有道2安装。
> 卡logo你可以试试按重装有道满分系统的教程重装看，或者去tg群里问

怎么不支持021，瞎讲
02X全系基本都一个样

**Reply by tofuture-1** [reply]
我印象中没刷满分版之前的版本号是ydp021，按教程是要刷了系统到版本号为2.0.7之类的才能装penmod，如果一样适配那抱歉了，是我没注意查证

**Reply by Redbeanw44602** [reply]
> 我印象中没刷满分版之前的版本号是ydp021，按教程是要刷了系统到版本号为2.0.7之类的才能装penmod，如果一样适配那抱歉了，是我没注意查证

你的观察没有错，但是很明显型号和系统版本没有什么关联.......
PenMods在某个版本后会自动修正021的系统sku到022（由于某些原因）

**Comment by OXiaoZhiO** [comment]
YDPA6-2,版本4.7.2行不行

**Reply by tofuture-1** [reply]
估计不行，项目的readme中有适用说明

**Comment by balckN** [comment]
===== RESTART: C:\Users\Administrator\Desktop\Installer-1.1.0\Installer.py =====
┌───────────────────────────────────┐
│                                   │
│       ╔═╗┌─┐┌┐┌╔╦╗┌─┐┌┬┐┌─┐       │
│       ╠═╝├┤ │││║║║│ │ ││└─┐       │
│       ╩  └─┘┘└┘╩ ╩└─┘─┴┘└─┘       │
│                                   │
└───────────────────────────────────┘
Welcome to use PenMods!
Developer:  RedbeanW
Repo:       https://github.com/PenUniverse/Installer
Version:    1.1.0
Traceback (most recent call last):
  File "D:\Lib\site-packages\urllib3\connectionpool.py", line 773, in urlopen
    self._prepare_proxy(conn)
  File "D:\Lib\site-packages\urllib3\connectionpool.py", line 1042, in _prepare_proxy
    conn.connect()
  File "D:\Lib\site-packages\urllib3\connection.py", line 721, in connect
    self._tunnel()
  File "D:\Lib\http\client.py", line 971, in _tunnel
    (version, code, message) = response._read_status()
  File "D:\Lib\http\client.py", line 300, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\Lib\site-packages\requests\adapters.py", line 667, in send
    resp = conn.urlopen(
  File "D:\Lib\site-packages\urllib3\connectionpool.py", line 841, in urlopen
    retries = retries.increment(
  File "D:\Lib\site-packages\urllib3\util\retry.py", line 474, in increment
    raise reraise(type(error), error, _stacktrace)
  File "D:\Lib\site-packages\urllib3\util\util.py", line 38, in reraise
    raise value.with_traceback(tb)
  File "D:\Lib\site-packages\urllib3\connectionpool.py", line 773, in urlopen
    self._prepare_proxy(conn)
  File "D:\Lib\site-packages\urllib3\connectionpool.py", line 1042, in _prepare_proxy
    conn.connect()
  File "D:\Lib\site-packages\urllib3\connection.py", line 721, in connect
    self._tunnel()
  File "D:\Lib\http\client.py", line 971, in _tunnel
    (version, code, message) = response._read_status()
  File "D:\Lib\http\client.py", line 300, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Administrator\Desktop\Installer-1.1.0\Installer.py", line 289, in <module>
    **handleAssert(checkConnection(), '安装程序需要联网执行，请检查您的网络链接。')**
  File "C:\Users\Administrator\Desktop\Installer-1.1.0\Installer.py", line 210, in checkConnection
    return requests.get(PENMODS_SERVER_ADDR).status_code == 200
  File "D:\Lib\site-packages\requests\api.py", line 73, in get
    return request("get", url, params=params, **kwargs)
  File "D:\Lib\site-packages\requests\api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "D:\Lib\site-packages\requests\sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "D:\Lib\site-packages\requests\sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "D:\Lib\site-packages\requests\adapters.py", line 682, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))





请问这种情况该怎么解决

**Reply by tofuture-1** [reply]
看路径你这个是原版的安装程序吧？目前无法正常使用，原因在本帖开始有说明。现在可以去tg群里手动装1.3.0预览版，或者拿本帖开篇的修改版程序给有道2安装1.2.0，使用前注意看清楚说明

**Comment by balckN** [comment]
感谢，当时眼瞎尽然没看见
________________________________
发件人: forward ***@***.***>
发送时间: 2025年4月26日 16:23
收件人: PenUniverse/PenMods-release ***@***.***>
抄送: zangxuyao ***@***.***>; Comment ***@***.***>
主题: Re: [PenUniverse/PenMods-release] penmod1.2.0安装脚本 (Discussion #177)


看路径你这个是原版的安装程序吧？目前无法正常使用，原因在本帖开始有说明。现在可以去tg群里手动装1.3.0预览版，或者拿本帖开篇的修改版程序给有道2安装1.2.0，使用前注意看清楚说明

―

**Comment by zero1749** [comment]
import os
# import requests
import time
import json
import zipfile
import shutil
import sys

PENMODS_SERVER_ADDR    = "https://dictpen.amd.rocks/"
PENMODS_PUBLIC_PACKS   = PENMODS_SERVER_ADDR + "public_packs"
PENMODS_MOD_PACKS      = PENMODS_SERVER_ADDR + "mod_packs"

MOD_PACK_VERSION    = 100
PUBLIC_PACK_VERSION = 100

VERSION = "1.1.0"

def printLogo():
    print("""┌───────────────────────────────────┐
│                                   │
│       ╔═╗┌─┐┌┐┌╔╦╗┌─┐┌┬┐┌─┐       │
│       ╠═╝├┤ │││║║║│ │ ││└─┐       │
│       ╩  └─┘┘└┘╩ ╩└─┘─┴┘└─┘       │
│                                   │
└───────────────────────────────────┘
Welcome to use PenMods!
Developer:  RedbeanW
Repo:       https://github.com/PenUniverse/Installer
Version:    %s""" % VERSION)

def terminate():
    os.system('pause')
    sys.exit(-1)

def handleAssert(express, errmsg:str=None):
    if express:
        return
    if errmsg:
        print('Assertion Failed: %s' % errmsg)
        return

def initDirs():
    os.makedirs('dependents', exist_ok=True)
    shutil.rmtree('temp', ignore_errors=True)
    os.makedirs('temp', exist_ok=True)

class ADB:

    first_connected_device = None

    def __init__(self) -> None:
        pass

    """ get installed adb command prefix """
    def getCommand(self) -> str:
        adb = 'dependents\\platform-tools\\adb.exe'
        return adb if os.path.exists(adb) else 'adb'

    """ check adb install """
    def check(self) -> bool:
        return Utils.run(self.getCommand()).find('Android Debug Bridge') != -1 \

    """ automatic bypass adb auth """
    def bypassVerification(self) -> bool:
        if not self.first_connected_device:
            return False
        if self._execute('shell ls').find('login with "adb shell auth" to continue') != -1:
            if Utils.run('echo CherryYoudao | %s shell auth' % self.getCommand()).find('success') != -1:
                print('DictPen is unlocked.')
                return True
            else:
                return False
        return True

    """ test current device """
    def test(self) -> bool:
        return self._execute('shell uname -a').find('Linux') != -1

    """ protected execute adb command """
    def execute(self, cmd):
        if not self.check():
            handleAssert(self.install(), 'ADB服务是必须的。')
        if not self.test():
            handleAssert(self.connect(), '请连接你的设备。')
        self.bypassVerification()
        return self._execute(cmd)

    """ only generate command, without protect """
    def _execute(self, cmd):
        return Utils.run('%s %s' % (self.getCommand(),cmd))

    """ automatic install adb service """
    def install(self) -> bool:
        if self.check():
            return True
        try:
            handleAssert(public_pack['adb'])
        except:
            print('无法获取软件包列表。')
            return False
        # handleAssert(Utils.download(public_pack['adb'],'temp/adb.zip'), '无法下载Platform-Tools，ADB安装失败。')
        # file = zipfile.ZipFile('temp/adb.zip')
        # shutil.rmtree('dependents/', ignore_errors=True)
        # file.extractall(path='dependents/')
        # if self.check():
        #     print('ADB安装成功。')
        #     return True
        # return False

    """ automatic connect adb device """
    def connect(self) -> bool:
        print('# 请将词典笔与电脑连接，并开启ADB调试。')
        print('# 请进入 [更多设置]->[关于]->[法律监管] 后快速连续点击5次屏幕即可打开ADB调试。')
        waitingTime = 0
        while True:
            waitingTime += 1
            if (waitingTime > 120):
                print('\n长时间未与设备建立连接，进程退出')
                return False
            devices = []
            preText = '正在等待'
            for line in Utils.run(self.getCommand() + ' devices').split('\n'):
                if (len(line) < 5 or line.find('List of devices attached') != -1):
                    continue
                dev = line.split('\t')
                devices.append((dev[0],dev[1].removesuffix('\n')))
            if len(devices) > 0:
                device = devices[0]
                if len(devices) > 1:
                    preText = '! 请不要连接多个设备'
                else:
                    if (device[1] != 'device'):
                        preText = '! 设备 %s 已连接，但设备状态不正确(%s)，请检查' % (device[0],device[1])
                    else:
                        if self.first_connected_device and device[0] != self.first_connected_device:
                            preText = '! 这是断线重连，请不要连接不同的设备'
                        else:
                            print('\n设备 %s 已连接' % device[0])
                            print('# 在安装程序运行过程中，请保持设备连接。')
                            self.first_connected_device = device[0]
                            return True
            print('\r%s...%ss' % (preText,waitingTime),end='')
            time.sleep(1)

class Utils:

    # def __init__(self) -> None:
    #     pass

    # @staticmethod
    # def download(url, path:str):  # from: https://blog.csdn.net/weixin_43347550/article/details/105248223
    #     start = time.time()
    #     response = requests.get(url, stream=True)
    #     size = 0
    #     chunk_size = 1024
    #     content_size = int(response.headers['content-length'])
    #     try:
    #         if response.status_code == 200:
    #             print('Downloading ... ({size:.2f} Mb).'.format(
    #                 size=content_size / chunk_size / 1024))
    #             with open(path, 'wb') as file:
    #                 for data in response.iter_content(chunk_size=chunk_size):
    #                     file.write(data)
    #                     size += len(data)
    #                     print('\r   '+'%s %.2f%%' % ('━' * int(size * 50 / content_size),
    #                         float(size / content_size * 100)), end=' ')
    #         else:
    #             print(url)
    #             print('Wrong response code(%s)' % response.status_code)
    #             return False
    #         end = time.time()
    #         print('\nCompleted, time cost %.2f second(s).' % (end - start))
    #         return True
    #     except:
    #         print("Fail to download %s!" % url)
    #         return False

    @staticmethod
    def run(cmd: str) -> str:
        with os.popen(cmd) as res:
            buffer = res._stream.buffer.read()
        try:
            return buffer.decode().strip()
        except UnicodeDecodeError:
            return buffer.decode('gbk').strip()

adb         = None
mod_pack    = None
public_pack = None

def getSystemVersions() -> dict:
    ret = {}
    exec = adb.execute('shell cat /Version')
    if exec.find('No such file') != -1:
        return {}
    for i in exec.split('\n'):
        kv = i.split(':')
        if len(kv) >= 2: ret[kv[0]] = kv[1].strip('\r').removeprefix(' ')
    return ret

def getPcbaVersion() -> str:
    exec = adb.execute('shell get_pcba_version')
    return exec if exec.find('not found') == -1 else None

def getAppMd5() -> str:
    exec = adb.execute('shell md5sum -b /oem/YoudaoDictPen/output/YoudaoDictPen')
    return exec.split(' ')[0].lower() if exec.find('No such file') == -1 else None

# def checkConnection() -> bool:
#     return requests.get(PENMODS_SERVER_ADDR).status_code == 200

def isInstalled() -> bool:
    exec = adb.execute('shell cat /usr/bin/runDictPen')
    return exec.find('try_inject') == -1

# def install_a2(info: dict):
def install_a2():
    # handleAssert(Utils.download(info['download'],'temp/pack.zip'))
    # zipfile.ZipFile('temp/pack.zip').extractall('temp/pack')
    
    def ro_check():
        if adb.execute('shell touch /TEST_PART_RW').find('Read-only file system'):
            adb.execute('shell mount -o remount,rw /')
            time.sleep(1)
        adb.execute('shell rm -f /TEST_PART_RW')

    def upload_ex(path,to,setperm=False):
        ro_check()
        adb.execute('push "temp/pack/%s" "%s"' % (path,to))
        if setperm:
            adb.execute('shell chmod +x "%s"' % to)
    
    print('(1/4) 正在安装自动注入器...')
    upload_ex('dependents/injector','/userdisk/Loader/injector',True)
    upload_ex('script/try_inject','/usr/bin/try_inject',True)
    adb.execute("shell \"sed -i '1i try_inject &' /usr/bin/runDictPen\"")

    print('(2/4) 正在更新动态链接库...')
    upload_ex('dependents/libm.so.6','/lib/libm.so.6')
    upload_ex('dependents/libstdc++.so.6','/usr/lib/libstdc++.so.6')

    print('(3/4) 正在安装 PenMods...')
    upload_ex('libPenMods.so','/userdisk/Loader')

    print('(4/4) 安装完成，正在重启...')
    adb.execute('shell safe_powerdown')
    print('# 如果没有意外，PenMods 已成功安装到您的词典笔上。')
    print('# Enjoy it!')
    terminate()

def install_b1(info: dict):
    handleAssert(Utils.download(info['download'],'temp/pack.zip'))
    zipfile.ZipFile('temp/pack.zip').extractall('temp/pack')
    
    def ro_check():
        if adb.execute('shell touch /TEST_PART_RW').find('Read-only file system'):
            adb.execute('shell mount -o remount,rw /')
            time.sleep(1)
        adb.execute('shell rm -f /TEST_PART_RW')

    def upload_ex(path,to,setperm=False):
        ro_check()
        adb.execute('push "temp/pack/%s" "%s"' % (path,to))
        if setperm:
            adb.execute('shell chmod +x "%s"' % to)
    
    print('(1/3) 正在安装自动注入器...')
    upload_ex('dependents/injector','/userdisk/Loader/injector',True)
    upload_ex('script/try_inject','/usr/bin/try_inject',True)
    adb.execute("shell \"sed -i '1i try_inject &' /usr/bin/runDictPen\"")

    print('(2/3) 正在安装 PenMods...')
    upload_ex('libPenMods.so','/userdisk/Loader')

    print('(3/3) 安装完成，正在重启...')
    adb.execute('shell safe_powerdown')
    print('# 如果没有意外，PenMods 已成功安装到您的词典笔上。')
    print('# Enjoy it!')
    terminate()

if __name__ == '__main__':
    
    printLogo()
    # initDirs()

    handleAssert(sys.platform == 'win32', '安装程序暂只支持Windows系统。')

    # Initialization.
    adb = ADB()
    # handleAssert(checkConnection(), '安装程序需要联网执行，请检查您的网络链接。')
    
    # mod_pack = json.loads(requests.get(PENMODS_MOD_PACKS).content)
    # public_pack = json.loads(requests.get(PENMODS_PUBLIC_PACKS).content)
    # handleAssert(mod_pack['version'] and public_pack['version'], '无法获取在线 ModPacks.')
    # handleAssert(mod_pack['version'] == MOD_PACK_VERSION, 'MOD列包表版本不匹配，请升级安装程序。')
    # handleAssert(public_pack['version'] == PUBLIC_PACK_VERSION, '公共包列表版本不匹配，请升级安装程序。')

    # mod_pack = mod_pack['packs']
    # public_pack = public_pack['packs']
    
    # Preparing for the Adaptation Check.
    sys_version = getSystemVersions()['Version']
    sys_pcba    = getPcbaVersion()
    sys_app_md5 = getAppMd5()
    if not sys_version or not sys_pcba or not sys_app_md5:
        print('无法获取系统信息，连接的也许不是词典笔？')
        terminate()
    
    handleAssert(isInstalled(), '当前系统已安装PenMods，请先还原系统再执行安装。')

    # Match the appropriate pack.
    # matched = None
    # for pack in mod_pack:
    #     adapt  = pack['adaptation']
    #     if sys_version in adapt['version'] and sys_pcba in adapt['pcba'] and sys_app_md5 in adapt['app']:
    #         matched = pack
    #         break
    
    # if not matched:
    #     print('无法为当前词典笔匹配合适的PenMods，可能其暂时不受支持。')
    #     print('Version: %s' % sys_version)
    #     print('Pcba:    %s' % sys_pcba)
    #     print('App:     %s' % sys_app_md5)
    #     terminate()
    # ver = matched['version']
    # print('已匹配到合适的PenMods (%s,V%s), 是否确认安装？' % (matched['name'],'.'.join(str(v) for v in matched['version'])))
    # print('# 务必仔细阅读 github.com/PenUniverse/Installer 仓库中的注意事项')
    # print('# 安装 PenMods 可能导致您失去有道官方保修')
    # print('# 使用 PenMods 造成的一切后果均由您本人承担，与项目作者没有任何关系')
    # print('# 特别注意: 不要为已修改过的系统执行安装，这是安装程序并非升级程序！！')

    if input('您是否已充分阅读、理解与接受以上告示，并决定开始安装？[y/N] ').lower() != 'y':
        terminate()
    install_a2()
    # if matched['install'] == 'a2':
    #     install_a2(matched)
    # elif matched['install'] == 'b1':
    #     install_b1(matched)
    # else:
    #     print('远端要求使用不受支持的安装方法，尝试更新安装程序？')

**Comment by mcxgjkh** [comment]
安装器目前出现了这种情况，非常奇怪，有大佬帮我看看吗）））😂
(1/4) 正在安装自动注入器...
adb: error: cannot stat 'temp/pack/dependents/injector': No such file or directory
adb: error: cannot stat 'temp/pack/script/try_inject': No such file or directory
(2/4) 正在更新动态链接库...
adb: error: cannot stat 'temp/pack/dependents/libm.so.6': No such file or directory
adb: error: cannot stat 'temp/pack/dependents/libstdc++.so.6': No such file or directory
(3/4) 正在安装 PenMods...
adb: error: cannot stat 'temp/pack/libPenMods.so': No such file or directory
(4/4) 安装完成，正在重启...

**Reply by tofuture-1** [reply]
用的是修改版安装器吗？
